### PR TITLE
Add ALiBi and dropout to SDP

### PR DESCRIPTION
**Pull request by shadeMe**
Also add the the `AttentionScorer` trait which is implemented by SDP.

### CODE REVIEW

File: src/layers/attention.rs [hunk 1]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
     MergedSplitAfter(QkvSplit),
 }
 
+/// ALiBi: Linear biases for attention [Press et al., 2022](https://arxiv.org/abs/2108.12409).
```
Comment (danieldk) [inline src/layers/attention.rs]: Question: do we want hyperlinks in the inline descriptions? Maybe they distract from glossing an overview page?

Screenshot 2024-01-15 at 08 35 49

If not, I propose:

```suggestion
/// Linear biases for attention (ALiBi)
```

File: src/layers/attention.rs [hunk 2]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+pub struct AttentionLinearBiases {
+    slopes: Tensor,
+    is_causal: bool,
+    is_inverted: bool,
+}
+
+impl AttentionLinearBiases {
+    pub fn new(n_attention_heads: usize, is_causal: bool, is_inverted: bool) -> Result<Self> {
```
Comment (danieldk) [inline src/layers/attention.rs:70]: rustdoc

File: src/layers/attention.rs [hunk 3]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+    ///
+    /// * n_attention_heads - Number of attention heads.
+    ///
+    /// Returns: Head slope tensor.
+    /// *Shape:* `(1, heads, 1, 1)`
+    #[allow(clippy::unnecessary_cast)]
+    fn calculate_slopes(n_attention_heads: usize) -> Result<Tensor> {
+        fn slopes_with_step(n_attention_heads: usize, step: usize) -> Result<Tensor> {
+            let ratio = 2.0f32.powf(-8.0f32 / n_attention_heads as f32);
+            let slope = (1..(1 + n_attention_heads))
```
Comment (danieldk) [inline src/layers/attention.rs]:
```suggestion
            let slope = (1..n_attention_heads + 1)
```

looks more canonical.

File: src/layers/attention.rs [hunk 4]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+
+    /// Calculate the linear bias tensor upto a given (key) sequence length.
+    ///
+    /// * seq_len - Maximum number of timesteps to calculate.
+    ///
+    /// Returns: Multi-headed linear bias tensor.
+    /// *Shape:* `(1, heads, seq_len, seq_len)` (non-causal) or
+    /// `(1, heads, 1, seq_len)` (causal)
+    fn calculate_biases(&self, seq_len: usize) -> Result<Tensor> {
+        let mut distances = match self.is_causal {
```
Comment (danieldk) [inline src/layers/attention.rs]: Use `if` for booleans, less ceremony.

File: src/layers/attention.rs [hunk 5]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+            true => Tensor::arange(1 - seq_len as i64, 1, &Device::Cpu)?,
+            false => {
+                let mut distance = Tensor::arange(0, seq_len as i64, &Device::Cpu)?;
+                distance = distance.broadcast_sub(&distance.reshape(((), 1))?)?;
+                distance
+                    .abs()?
+                    .broadcast_mul(&Tensor::new(-1i64, &Device::Cpu)?)?
```
Comment (danieldk) [inline src/layers/attention.rs]: Meh, no `i64` implementation of `Mul` in candle.

File: src/layers/attention.rs [hunk 6]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+    /// Apply linear biases to (unmasked) attention scores.
+    ///
+    /// * attention_scores - Attention scores.
+    ///   *Shape:* `(batch_size, heads, query_len, key_len)`
+    ///
+    /// Returns: Attention scores with linear biases.
+    /// *Shape:* `(batch_size, heads, query_len, key_len)`
+    pub fn forward(&self, attention_scores: &Tensor) -> Result<Tensor> {
+        let biases = self
+            .calculate_biases(attention_scores.dim(attention_scores.rank() - 1)?)?
```
Comment (danieldk) [inline src/layers/attention.rs]: Hmmm, I think in most cases I prefer doing

```
let (_, _, _, key_len) = attention_scores.shape().dims4()?;
```

before. Makes it easier to see what dimensionality is used, since it's named.

File: src/layers/attention.rs [hunk 7]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+            .to_dtype(attention_scores.dtype())?
+            .to_device(attention_scores.device())?;
```
Comment (danieldk) [inline src/layers/attention.rs:167]: We did it like this in curated, but is it more efficient to do the device placement at the last moment? (I guess so, since we'd have many runs of computationally small kernels otherwise.)

File: src/layers/attention.rs [hunk 8]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+
+        Ok(attention_scores.add(&biases)?)
+    }
+}
+
+pub trait AttentionScorer {
```
Comment (danieldk) [inline src/layers/attention.rs:174]: rustdoc

File: src/layers/attention.rs [hunk 9]
```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
...
+    fn forward(
```
Comment (danieldk) [inline src/layers/attention.rs:190]: rustdoc

File: src/layers/attention.rs [hunk 10]
```diff
@@ -72,11 +199,17 @@ impl ScaledDotProductAttention {
...
-        attn_scores = attention_mask.apply_logit_mask(&attn_scores)?;
         let temperature = (model_width as f64).sqrt();
-        let attn_scores = (attn_scores / temperature)?;
+        attn_scores = (attn_scores / temperature)?;
+
+        if let Some(linear_biases) = &self.linear_biases {
+            attn_scores = linear_biases.forward(&attn_scores)?;
+        }
+        attn_scores = attention_mask.apply_logit_mask(&attn_scores)?;
+
```
Comment (danieldk) [inline src/layers/attention.rs:233]:
```suggestion

        attn_scores = attention_mask.apply_logit_mask(&attn_scores)?;
```

To make it clearer that applying the logit mask is related to softmax.

**Review by danieldk — COMMENTED** [review]
Awesome, left a bunch of minor comments, feel free to merge afterwards.

File: src/layers/attention.rs (hunk 1 above)
Comment (shadeMe) [reply src/layers/attention.rs]: I'd like to keep the links to the papers in some way, if possible. Maybe we can stick to the format you're currently using in your transformer layer PR (`See XZY et al. (2001)`)?

File: src/layers/attention.rs (hunk 5 above)
Comment (shadeMe) [reply src/layers/attention.rs]: As in no vectorized implementation? Does it get coerced to another type internally (the tests appear to be working)?

File: src/layers/attention.rs (hunk 1 above)
Comment (danieldk) [reply src/layers/attention.rs]: Yeah, I think having a link in the detailed description is great to have.

File: src/layers/attention.rs (hunk 5 above)
Comment (danieldk) [reply src/layers/attention.rs]: I was more lamenting that there is no implementation for `Tensor` of `Mul` taking i64 (only floating point types). Would be nice if candle had that, because then we could just say `distance.abs()? * -1i64`.

File: src/layers/attention.rs (hunk 7 above)
Comment (shadeMe) [reply src/layers/attention.rs:167]: Yeah, I'd think so too? Don't think we actually profiled it at this level.

File: src/layers/attention.rs (hunk 2 above)
Comment (shadeMe) [reply src/layers/attention.rs:70]: Added.

File: src/layers/attention.rs (hunk 5 above)
Comment (shadeMe) [reply src/layers/attention.rs]: Ah, right. Yeah, having to manually allocate `Tensor`s for non-fp scalars is indeed meh.

File: src/layers/attention.rs (hunk 6 above)
Comment (shadeMe) [reply src/layers/attention.rs]: Fixed.
